sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/BeamSqlFnExecutor.java

```diff
@@ -503,13 +513,32 @@ private static boolean isDateNode(SqlTypeName type, Object value) {
   public void prepare() {}
 
   @Override
-  public List<Object> execute(
-      Row inputRow, BoundedWindow window, ImmutableMap<Integer, Object> correlateEnv) {
-    List<Object> results = new ArrayList<>();
-    for (BeamSqlExpression exp : exps) {
-      results.add(exp.evaluate(inputRow, window, correlateEnv).getValue());
+  public @Nullable List<Object> execute(
+      Row inputRow, BoundedWindow window, BeamSqlExpressionEnvironment env) {
+
+    final BeamSqlExpressionEnvironment finalEnv = buildExprEnv(inputRow, window, env);
+
+    boolean conditionResult = filterCondition.evaluate(inputRow, window, finalEnv).getBoolean();
+
+    if (conditionResult) {
+      return projections
+          .stream()
+          .map(project -> project.evaluate(inputRow, window, finalEnv).getValue())
+          .collect(Collectors.toList());
+    } else {
+      return null;
+    }
+  }
+
+  private BeamSqlExpressionEnvironment buildExprEnv(
+      Row inputRow, BoundedWindow window, BeamSqlExpressionEnvironment env) {
+    int nextExprId = 0;
+    BeamSqlExpressionEnvironment localEnv = env.copyWithEmptyLocalRefs(exprs.size());
+    for (BeamSqlExpression expr : exprs) {
+      localEnv.setLocalRef(nextExprId, expr.evaluate(inputRow, window, localEnv));
```

Comment:
should the evaluation be lazy performed at access time?

Reply:
That is a good question. I'm not actually sure. I am going to try to get the other things green here and then consider that change.

Reply:
This is actually a really helpful comment, because: right now the environment is coded as though it can be re-used for multiple rows and windows (just because it doesn't reference them directly). But the values set in the env are tightly coupled to the current row and window, so it isn't actually independent. To do the environment lazily they need to be baked in, so that actually clarifies and simplifies more.